### PR TITLE
✨ Add DryRunClient wrapper

### PR DESCRIPTION
**Pull request by alvaroaleman**
This PR adds a DryRunClient wrapper that enforced DryRun on all mutating operations and plugs that into the client and manager constructors. This is useful to check what a controller does in a production environment without actually doing it.

Fixes https://github.com/kubernetes-sigs/controller-runtime/issues/831

### CODE REVIEW

File: pkg/client/client_test.go [hunk 1]
```diff
@@ -33,9 +33,9 @@ import (
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/runtime/schema"
-	"sigs.k8s.io/controller-runtime/pkg/client"
-
 	kscheme "k8s.io/client-go/kubernetes/scheme"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
```
Comment (djzager) [inline pkg/client/client_test.go]: Only curious because it looks intentional and not the kind of thing I would expect `gofmt` to do. Why this change?

File: pkg/client/dryrun_test.go [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+
+		opts := &client.PatchOptions{DryRun: []string{"Bye", "Pippa"}}
+
+		Expect(getClient().Status().Patch(ctx, changedDep, client.MergeFrom(dep), opts)).ToNot(HaveOccurred())
+
+		actual, err := clientset.AppsV1().Deployments(ns).Get(dep.Name, metav1.GetOptions{})
+		Expect(err).NotTo(HaveOccurred())
+		Expect(actual).NotTo(BeNil())
+		Expect(actual).To(BeEquivalentTo(dep))
+	})
```
Comment (djzager) [inline pkg/client/dryrun_test.go]: Reading https://kubernetes.io/blog/2019/01/14/apiserver-dry-run-and-kubectl-diff/ I see that one of the advantages of supporting a dry-run, besides not being persisted, is

> the request is still processed as typical request: the fields are defaulted, the object is validated, it goes through the validation admission chain, and through the mutating admission chain, and then the final object is returned to the user as it normally would, without being persisted.

I think a test case where a create or an update failed validation would be a good addition.

File: pkg/client/client_test.go (hunk 1 above)
Comment (alvaroaleman) [reply pkg/client/client_test.go]: Because we usually put imports form our own repo in a dedicated group, rather than mixing them with other imports in a common group

File: pkg/client/dryrun_test.go [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+		Expect(apierrors.IsNotFound(err)).To(BeTrue())
+	})
+
+	It("should refuse a create request for an invalid object", func() {
+		changedDep := dep.DeepCopy()
+		changedDep.Spec.Template.Spec.Containers = nil
+
+		err := getClient().Create(ctx, changedDep)
+		Expect(apierrors.IsInvalid(err)).To(BeTrue())
+
```
Comment (djzager) [inline pkg/client/dryrun_test.go]: nit for consistency.

```suggestion
```

File: pkg/client/dryrun_test.go [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+		Expect(actual).To(BeEquivalentTo(dep))
+	})
+
+	It("should refuse an update request for an invalid change", func() {
+		changedDep := dep.DeepCopy()
+		changedDep.Spec.Template.Spec.Containers = nil
+
+		err := getClient().Update(ctx, changedDep)
+		Expect(apierrors.IsInvalid(err)).To(BeTrue())
+
```
Comment (djzager) [inline pkg/client/dryrun_test.go]: nit for consistency.

```suggestion
```

File: pkg/client/client.go [hunk 5]
```diff
@@ -37,6 +37,9 @@ type Options struct {
 
 	// Mapper, if provided, will be used to map GroupVersionKinds to Resources
 	Mapper meta.RESTMapper
+
+	// DryRun specifies whether dry run mode should be enforced
+	DryRun bool
```
Comment (DirectXMan12) [inline pkg/client/client.go]: What's the main usecase for having this as an option here vs just letting people call `NewDryRunClient`?  Just want to avoid too many options here, esp since this isn't strictly related to creating the client

Comment (alvaroaleman) [reply pkg/client/client.go]: Good point, its not really necessary there, I removed it.

I would like to keep it in the `manager` though, because the `defaultNewClient` is not exported and replacing the `NewClient` func is not very intuitive and an option ppl that are not deeply familiar with c-r will most likely not become aware of. WDYT?

**Comment by alvaroaleman** [comment]
/assign @DirectXMan12 
Could you have another look, please?

**Comment by alvaroaleman** [comment]
@DirectXMan12 any chance you can have a look?

File: pkg/client/client.go (hunk 5 above)
Comment (mengqiy) [reply pkg/client/client.go]: Not super excited about growing manager's Options.

One alternative we can try here is that:
Making `defaultNewClient` exportable, but under `pkg/manager/internal` or `pkg/internal/manager` or some place similar.
Then you can have a helper method that satisfy `NewClientFunc`
```
func NewDryRunClientFn(cache cache.Cache, config *rest.Config, options client.Options) (client.Client, error) {
  cl, err := managerinternal.DefaultNewClient(cache, config, options) 
  return &dryRunClient{cl}, err
}
```
User can pass this method to manager Options.

@DirectXMan12 Thoughts?

Comment (alvaroaleman) [reply pkg/client/client.go]: @mengqiy the reason I dislike that is that its pretty hard to discover that these kind of build-in opts exists

**Comment by DirectXMan12** [comment]
hey, sorry for the lag

/lgtm
/approve

**Comment by alvaroaleman** [comment]
@DirectXMan12 could you give a new lgtm as I had to rebase?

**Comment by DirectXMan12** [comment]
FWIW, I'd love to eventually figure out how to make it easy to sub in the dryrun client instead of needing an option on manager, but I don't really think I want to block on that now
